base/stats.go

```diff
@@ -468,43 +464,47 @@ func (s *SgwFloatStat) Describe(ch chan<- *prometheus.Desc) {
 }
 
 func (s *SgwFloatStat) Collect(ch chan<- prometheus.Metric) {
-	ch <- prometheus.MustNewConstMetric(s.statDesc, s.statValueType, s.Val, getValues(s.labels)...)
+	ch <- prometheus.MustNewConstMetric(s.statDesc, s.statValueType, math.Float64frombits(atomic.LoadUint64(&s.Val)), getValues(s.labels)...)
 }
 
 func (s *SgwFloatStat) Set(newV float64) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	s.Val = newV
+	atomic.StoreUint64(&s.Val, math.Float64bits(newV))
 }
 
 func (s *SgwFloatStat) SetIfMax(newV float64) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	if newV > s.Val {
-		s.Val = newV
+	for {
+		cur := atomic.LoadUint64(&s.Val)
+		curVal := math.Float64frombits(cur)
+		nxtVal := curVal + newV
+		nxt := math.Float64bits(nxtVal)
+		if atomic.CompareAndSwapUint64(&s.Val, cur, nxt) {
+			return
+		}
 	}
 }
 
-func (s *SgwFloatStat) Add(newV float64) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	s.Val += newV
+func (s *SgwFloatStat) Add(delta float64) {
+	for {
+		cur := atomic.LoadUint64(&s.Val)
+		curVal := math.Float64frombits(cur)
+		nxtVal := curVal + delta
+		nxt := math.Float64bits(nxtVal)
+		if atomic.CompareAndSwapUint64(&s.Val, cur, nxt) {
+			return
+		}
+	}
 }
 
 func (s *SgwFloatStat) MarshalJSON() ([]byte, error) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	return []byte(fmt.Sprintf("%v", s.Val)), nil
+	return []byte(fmt.Sprintf("%v", math.Float64frombits(atomic.LoadUint64(&s.Val)))), nil
 }
 
 func (s *SgwFloatStat) String() string {
 	return fmt.Sprintf("%v", s.Val)
 }
```

Comment:
Use FormatFloat for these for similar performance reasons. Need to also consider what format the float is represented in, and whether that's valid in the JSON number format.